File: 5. Estructuras de datos no lineales/JavaScript/arboles_binarios.js
// 1. Crea un árbol binario e implementa un método para recorrerlo en preorden, inorden y postorden.
// 2. Escribe una función para buscar un elemento específico en un árbol binario.

class Nodo {
  constructor(valor) {
    this.valor = valor;
    this.izquierda = null;
    this.derecha = null;
  }
}

class ArbolBinario {
  constructor() {
    this.raiz = null;
  }

  insertar(valor) {
    const nuevoNodo = new Nodo(valor);

    if (!this.raiz) {
      this.raiz = nuevoNodo;
    } else {
      this.insertarNodo(this.raiz, nuevoNodo);
    }
  }

  insertarNodo(nodo, nuevoNodo) {
    if (nuevoNodo.valor < nodo.valor) {
      if (!nodo.izquierda) {
        nodo.izquierda = nuevoNodo;
      } else {
        this.insertarNodo(nodo.izquierda, nuevoNodo);
      }
    } else {
      if (!nodo.derecha) {
        nodo.derecha = nuevoNodo;
      } else {
        this.insertarNodo(nodo.derecha, nuevoNodo);
      }
    }
  }

  preorden(nodo = this.raiz) {
    if (!nodo) return;
    console.log(nodo.valor);
    this.preorden(nodo.izquierda);
    this.preorden(nodo.derecha);
  }

  inorden(nodo = this.raiz) {
    if (!nodo) return;
    this.inorden(nodo.izquierda);
    console.log(nodo.valor);
    this.inorden(nodo.derecha);
  }

  postorden(nodo = this.raiz) {
    if (!nodo) return;
    this.postorden(nodo.izquierda);
    this.postorden(nodo.derecha);
    console.log(nodo.valor);
  }

  buscar(valor, nodo = this.raiz) {
    if (!nodo) return false;
    if (nodo.valor === valor) return true;
    if (valor < nodo.valor) {
      return this.buscar(valor, nodo.izquierda);
    } else {
      return this.buscar(valor, nodo.derecha);
    }
  }
}

const arbol = new ArbolBinario();
arbol.insertar(10);
arbol.insertar(5);
arbol.insertar(15);
arbol.insertar(3);
arbol.insertar(7);
arbol.insertar(12);
arbol.insertar(17);

console.log("Preorden:");
arbol.preorden();

console.log("Inorden:");
arbol.inorden();

console.log("Postorden:");
arbol.postorden();

console.log("Buscar 7:", arbol.buscar(7));
console.log("Buscar 8:", arbol.buscar(8));

File: 5. Estructuras de datos no lineales/Python/arboles_binarios.py
# 1. Implementa una clase para un árbol binario que permita agregar nodos y realizar un recorrido en orden.
# 2. Escribe una función para calcular la altura de un árbol binario.


class Nodo:
    def __init__(self, valor):
        self.valor = valor
        self.izquierda = None
        self.derecha = None


class ArbolBinario:
    def __init__(self):
        self.raiz = None

    def agregar(self, valor):
        if self.raiz is None:
            self.raiz = Nodo(valor)
        else:
            self._agregar(valor, self.raiz)

    def _agregar(self, valor, nodo):
        if valor < nodo.valor:
            if nodo.izquierda is None:
                nodo.izquierda = Nodo(valor)
            else:
                self._agregar(valor, nodo.izquierda)
        else:
            if nodo.derecha is None:
                nodo.derecha = Nodo(valor)
            else:
                self._agregar(valor, nodo.derecha)

    def inorden(self):
        self._inorden(self.raiz)

    def _inorden(self, nodo):
        if nodo is not None:
            self._inorden(nodo.izquierda)
            print(nodo.valor)
            self._inorden(nodo.derecha)

    def altura(self):
        return self._altura(self.raiz)

    def _altura(self, nodo):
        if nodo is None:
            return 0
        else:
            return 1 + max(self._altura(nodo.izquierda), self._altura(nodo.derecha))


arbol = ArbolBinario()
arbol.agregar(5)
arbol.agregar(3)
arbol.agregar(7)
arbol.agregar(2)
arbol.agregar(4)
arbol.agregar(6)
arbol.agregar(8)

print("Inorden:")
arbol.inorden()

print("Altura:")
print(arbol.altura())
